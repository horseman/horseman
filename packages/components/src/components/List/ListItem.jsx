import styled, { css } from "styled-components";
import PropTypes from "prop-types";

import media from "../../styleUtils/media";

/**
 * Calculate the width based on the mobile base and the gutter
 */
const getItemWidth = (mobileBase, gutter) => {
  if (mobileBase > 1) {
    return `calc(${100 / mobileBase}% - ${gutter / 2}px)`;
  }
  return `${100 / mobileBase}%`;
};

const ListItem = styled.div`
  width: ${({ mobileBase, gutter }) => getItemWidth(mobileBase, gutter)};
  margin-top: ${({ gutter }) => gutter}px;

  &:nth-child(-n + ${props => props.mobileBase}) {
    margin-top: 0;
  }

  ${({ breakpoint }) =>
    media[breakpoint](css`
      width: calc(${({ base, gutter }) => `${100 / base}% - ${gutter / 2}px`});
      &:nth-child(-n + ${({ base }) => base}) {
        margin-top: 0em;
      }
    `)};
`;

ListItem.defaultProps = {
  gutter: 30,
  mobileBase: 1,
  breakpoint: "mdAndUp",
};

ListItem.propTypes = {
  base: PropTypes.number.isRequired,
  mobileBase: PropTypes.number,

  /**
   * The breakpoint at which the layout will go desktop to mobile
   */
  breakpoint: PropTypes.string,
};

export default ListItem;
