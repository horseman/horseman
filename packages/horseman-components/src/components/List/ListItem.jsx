import styled, { css } from "styled-components";
import PropTypes from "prop-types";

import media from "../../styleUtils/media";

/**
 * Calculate the width based on the mobile base and the gutter
 */
const getItemWidth = (mobileBase, gutter) => {
  if (mobileBase > 1) {
    return `calc(${100 / mobileBase}% - ${gutter / 2}px)`;
  }
  return `${100 / mobileBase}%`;
};

const ListItem = styled.div`
  width: ${({ mobileBase, gutter }) => getItemWidth(mobileBase, gutter)};
  margin-top: ${({ gutter }) => gutter}px;

  &:nth-child(-n + ${props => props.mobileBase}) {
    margin-top: 0;
  }

  ${media.mdAndUp(css`
    width: calc(${({ base, gutter }) => `${100 / base}% - ${gutter / 2}px`});
    &:nth-child(-n + ${({ base }) => base}) {
      margin-top: 0em;
    }
  `)};
`;

ListItem.defaultProps = {
  gutter: 30,
  mobileBase: 1,
};

ListItem.propTypes = {
  base: PropTypes.number.isRequired,
  mobileBase: PropTypes.number,
};

export default ListItem;
