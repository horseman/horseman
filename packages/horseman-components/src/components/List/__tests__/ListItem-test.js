import React from "react";
import { shallow } from "enzyme";

import ListItem from "../ListItem";

describe("ListItem", () => {
  test("renders with a base and default mobileBase", () => {
    const wrapper = shallow(<ListItem base={2} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("renders correctly with a mobileBase > 1", () => {
    const wrapper = shallow(<ListItem base={2} mobileBase={2} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("renders correctly with custom breakpoint", () => {
    const wrapper = shallow(
      <ListItem base={2} mobileBase={2} breakpoint="smAndUp" />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
