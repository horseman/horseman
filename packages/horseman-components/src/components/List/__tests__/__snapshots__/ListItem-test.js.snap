// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders correctly with a mobileBase > 1 1`] = `
.c0 {
  width: calc(50% - NaNpx);
  margin-top: 30px;
}

.c0:nth-child(-n + 2) {
  margin-top: 0;
}

@media (min-width:892px) {
  .c0 {
    width: calc(50% - 15px);
  }

  .c0:nth-child(-n + 2) {
    margin-top: 0em;
  }
}

<div
  className="c0"
/>
`;

exports[`ListItem renders with a base and default mobileBase 1`] = `
.c0 {
  width: 100%;
  margin-top: 30px;
}

.c0:nth-child(-n + 1) {
  margin-top: 0;
}

@media (min-width:892px) {
  .c0 {
    width: calc(50% - 15px);
  }

  .c0:nth-child(-n + 2) {
    margin-top: 0em;
  }
}

<div
  className="c0"
/>
`;
